serving/serving.go

```diff
@@ -82,13 +81,170 @@ func (serv *FeatureServer) TrainingData(req *pb.TrainingDataRequest, stream pb.F
 	return nil
 }
 
+func (serv *FeatureServer) TrainingTestSplit(stream pb.Feature_TrainingTestSplitServer) error {
+
+	var (
+		trainIter, testIter provider.TrainingSetIterator
+		isTrainFinished     bool
+		isTestFinished      bool
+	)
+
+	for {
+		if isTestFinished && isTrainFinished {
+			// If both iterators are finished, we can close the stream
+			serv.Logger.Infow("Both iterators are finished, closing stream")
+			return nil
+		}
+
+		req, err := stream.Recv()
+		if err != nil {
+			return err
+		}
+
+		id := req.GetId()
+		name, variant := id.GetName(), id.GetVersion()
+		logger := serv.Logger.With("Name", name, "Variant", variant, "RequestType", req.GetRequestType().String())
+
+		logger.Infow("Getting training test split", "Request", req.String())
+
+		featureObserver := serv.Metrics.BeginObservingTrainingServe(name, variant)
+		defer featureObserver.Finish()
+
+		switch req.GetRequestType() {
+		case pb.RequestType_INITIALIZE:
+			if err := serv.handleSplitInitializeRequest(stream, req, &trainIter, &testIter, logger); err != nil {
+				featureObserver.SetError()
+				return err
+			}
+		default:
+			if err := serv.handleSplitDataRequest(stream, req, &trainIter, &testIter, &isTestFinished, &isTrainFinished, logger); err != nil {
+				featureObserver.SetError()
+				return err
+			}
+		}
+	}
+}
+
+func (serv *FeatureServer) handleSplitInitializeRequest(
+	stream pb.Feature_TrainingTestSplitServer,
+	req *pb.TrainingTestSplitRequest,
+	trainIterator *provider.TrainingSetIterator,
+	testIterator *provider.TrainingSetIterator,
+	logger *zap.SugaredLogger,
+) error {
+	logger.Infow("Initializing dataset", "id", req.Id, "shuffle", req.Shuffle, "testSize", req.TestSize)
+
+	train, test, dropViews, err := serv.getTrainingSetTestSplitIterator(
+		req.Id.GetName(),
+		req.Id.GetVersion(),
+		req.TestSize,
+		req.Shuffle,
+		int(req.RandomState),
+	)
+
+	if err != nil {
+		logger.Errorw("Failed to get training set iterator", "Error", err)
+		return err
+	}
+	defer dropViews()
```

Comment:
not totally convinced I need this yet:

What i was thinking

dataset = get_training_set(ts.name, ts.var)
train, test = dataset.training_test_split(random_state=0)

# if you don't drop the views after these are consumed, its extra stuff in their provider -- which really isn't a big deal -- and also running dataset.training_test_split(random_state=0) won't be a new random set

I could probably get rid of the view dropping